awscli/customizations/paginate.py

```diff
@@ -161,11 +163,27 @@ def unify_paging_params(argument_table, operation_model, event_name,
                                      parse_type=type_name,
                                      serialized_name='MaxItems'),
                         shadowed_args)
+    # We will register two pagination handlers.
+    # 
+    # The first is focused on analyzing the CLI arguments passed to see
+    # if they contain explicit operation-specific pagination args.  If so,
+    # they are doing manual pagination and we turn off the CLI pagination.
+    # 
+    # The second is call later in the event chain and analyzes the actual
+    # calling parameters passed to the operation.
+    #
+    # The reason we have to do the second is that someone could put
+    # operation-specific pagination args in the CLI input JSON file
+    # directly and this bypasses all of the CLI args processing.
     session.register(
         parsed_args_event,
         partial(check_should_enable_pagination,
                 list(_get_all_cli_input_tokens(paginator_config)),
                 shadowed_args, argument_table))
+    session.register(
```

Comment:
I wonder if it makes sense to move this handler's logic to the `CliInputJSONArgument.add_to_call_parameters` handler and have `CliInputJSONArgument.add_to_call_parameters` turn off the `parsed_globals.paginate` flag? It looks like the `CliInputJSONArgument.add_to_call_parameters` and `check_should_enable_pagination_call_parameters` are both registered to the `calling-command` event, and we can isolate this logic to only when `--cli-input-json` parameter is used and so it does not need to be called for every paginatable command.

I do think it would make more sense to keep it in the `unify_paging_params` handler though if there are other instances where we set the `call_commands` and need to take account for account for pagination, but I do not believe there is right now. Thoughts?

Reply:
If I understand correctly, you are suggesting that we move this code to the `add_to_call_parameters` method of the `CLIInputJSONArgument` class.  You are correct that this handler is already being registered on the `calling-command` event we need.

I think the only downside I see to that is that the handler in `cliinputjson.py` would still need to know what pagination input tokens it should be looking for.  So, we would need some pagination-specific functionality in the code processing input JSON and that code would kind of duplicate what is already in `paginate.py`.

I think I have a slight preference to keep all of the pagination-related code in `paginate.py` but not for any really important reason.

Reply:
That's fair. I'm fine for keeping it in `paginate.py`